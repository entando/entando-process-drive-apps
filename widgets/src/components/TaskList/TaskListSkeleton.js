import React from 'react';
import PropTypes from 'prop-types';
import Skeleton from '@material-ui/lab/Skeleton';
import withStyles from '@material-ui/core/styles/withStyles';

const styles = {
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    padding: '10px 20px',
    '&>div': {
      width: '50%',
    },
  },
  search: {
    width: '20% !important',
    marginTop: 20,
  },
  tableRows: {
    margin: '40px auto 40px auto',
  },
};

const getKey = (qty, i) => qty - i;

const TaskListSkeleton = ({ classes, rows }) => (
  <>
    <div className={classes.header}>
      <div>
        <Skeleton width="40%" />
        <Skeleton width="20%" />
      </div>
      <div className={classes.search}>
        <Skeleton width="100%" height={40} variant="rect" />
      </div>
    </div>
    <div>
      <Skeleton width="100%" height={40} variant="rect" />
      {[...Array(rows)].map((row, i) => (
        <Skeleton key={getKey(rows, i)} className={classes.tableRows} width="98%" height={10} />
      ))}
    </div>
    <div>
      <Skeleton width="100%" height={50} variant="rect" />
    </div>
  </>
);

TaskListSkeleton.propTypes = {
  classes: PropTypes.shape({
    header: PropTypes.string,
    search: PropTypes.string,
    tableRows: PropTypes.string,
  }),
  rows: PropTypes.number,
};

TaskListSkeleton.defaultProps = {
  classes: {},
  rows: 10,
};

export default withStyles(styles)(TaskListSkeleton);
