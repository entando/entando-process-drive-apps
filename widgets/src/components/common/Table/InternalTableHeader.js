import PropTypes from 'prop-types';
import React from 'react';
import TableCell from '@material-ui/core/TableCell';

import columnType from '../../../types/columnType';
import TableSortLabel from './TableSortLabel';
import withStyles from '@material-ui/core/styles/withStyles';
import TableBulkSelectContext from './TableBulkSelectContext';
import { StyledCheckbox } from './custom/CheckboxCell';

const StyledHeaderCell = withStyles(
  {
    head: {
      backgroundColor: '#e5e6e7',
      color: '#323233',
      fontWeight: 'bold',
      borderBottom: '1px solid #e7eaec',
      padding: '9px 16px 10px',
    },
  },
  { name: 'StyledHeaderCell' }
)(TableCell);

function InternalTableHeader({ column, createSortHandler, sortedColumn, sortOrder }) {
  const check = column.header === '_checkbox' && (
    <TableBulkSelectContext.Consumer>
      {({ onSelectAll, onSelectNone }) => {
        const handleChecked = ({ target }) => (target.checked ? onSelectAll() : onSelectNone());
        return <StyledCheckbox onChange={handleChecked} />;
      }}
    </TableBulkSelectContext.Consumer>
  );
  const label = check || column.header || column.accessor;
  const header = column.sortFunction ? (
    <TableSortLabel
      key={column.accessor}
      active={column.accessor === sortedColumn}
      onClick={createSortHandler(column.accessor)}
      direction={sortOrder}
    >
      {label}
    </TableSortLabel>
  ) : (
    label
  );

  return (
    <StyledHeaderCell
      variant="head"
      key={column.accessor}
      align={column.align}
      style={{ whiteSpace: 'nowrap', ...column.styles }}
    >
      {header}
    </StyledHeaderCell>
  );
}

InternalTableHeader.propTypes = {
  column: columnType.isRequired,
  sortOrder: PropTypes.string.isRequired,
  createSortHandler: PropTypes.func,
  sortedColumn: PropTypes.string,
};

InternalTableHeader.defaultProps = {
  createSortHandler: () => () => {},
  sortedColumn: '',
};

export default InternalTableHeader;
