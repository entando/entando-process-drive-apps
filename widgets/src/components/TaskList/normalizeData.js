import ActionCell from 'components/common/Table/custom/ActionCell';
import { compareDates, compareNumbers, compareStrings } from 'components/common/Table/utils';

export const getType = (column, firstRow) => {
  let sortFunction = compareStrings;
  if (firstRow[column] instanceof Date) {
    sortFunction = compareDates;
  } else if (typeof firstRow[column] === 'number') {
    sortFunction = compareNumbers;
  }

  return sortFunction;
};

export const normalizeColumns = (columns, firstRow) => {
  const normalized = columns
    .filter(column => column.isVisible)
    .map(column => ({
      header: column.header,
      accessor: column.name,
      position: column.position,
      sortFunction: getType(column, firstRow),
    }));
  // order columns
  normalized.sort((a, b) => (a.position > b.position ? 1 : a.position < b.position ? -1 : 0));

  // add action field
  normalized.unshift({
    header: 'Action',
    accessor: 'action',
    customCell: ActionCell,
    styles: {
      position: 'sticky',
      left: 0,
      background: 'white',
      width: 20,
      zIndex: 100,
      borderRight: '1px solid #eee',
      paddingRight: 16,
      textAlign: 'center',
    },
  });

  return normalized;
};

export const normalizeRows = rows =>
  rows.map(row => {
    const normalizedRow = {};
    Object.keys(row).forEach(key => {
      if (row[key] instanceof Object) {
        normalizedRow[key] = '';
      } else {
        normalizedRow[key] = String(row[key]);
      }
    });
    return normalizedRow;
  });
