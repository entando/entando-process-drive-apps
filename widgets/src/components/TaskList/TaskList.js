import { ThemeProvider } from '@material-ui/styles';
import React from 'react';
import i18next from 'i18next';
import PropTypes from 'prop-types';
import Paper from '@material-ui/core/Paper';
import withStyles from '@material-ui/styles/withStyles';

import { WIDGET_CODES } from 'api/constants';
import configApi from 'api/config';
import taskListApi from 'api/taskList';
import utils from 'utils';

import { normalizeColumns, normalizeRows } from 'components/TaskList/normalizeData';
import ErrorNotification from 'components/common/ErrorNotification';
import ErrorComponent from 'components/common/ErrorComponent';
import Table from 'components/common/Table/Table';
import theme from 'theme';

const styles = {
  paper: {
    minHeight: 459,
    position: 'relative',
  },
};

class TaskList extends React.Component {
  state = {
    loading: true,
    columns: [],
    rows: [],
    size: 0,
    currentPage: 0,
    connection: {},
    error: null,
    errorAlert: null,
  };

  timer = { ref: null };

  componentDidMount = async () => {
    const { lazyLoading } = this.props;
    const { currentPage } = this.state;

    try {
      const config = await configApi.get(WIDGET_CODES.taskList);
      console.log(config);

      const taskList = lazyLoading
        ? await taskListApi.get(config.connection, currentPage, 10)
        : await taskListApi.get(config.connection);
      console.log(taskList);

      this.setState({
        loading: false,
        columns: normalizeColumns(config.columns, taskList.payload[0]),
        rows: normalizeRows(taskList.payload),
        size: taskList.metadata.size,
        connection: config.connection,
      });
    } catch (error) {
      this.handleError(error);
    }
  };

  componentDidUpdate = prevProps => {
    const { lazyLoading } = this.props;
    if (prevProps.lazyLoading !== lazyLoading) {
      this.updateRows(lazyLoading ? 0 : undefined);
    }
  };

  updateRows = async (
    page,
    rowsPerPage = 10,
    sortedColumn,
    sortedOrder,
    filter,
    callback = () => {},
    withDelay
  ) => {
    const { connection } = this.state;

    if (withDelay) {
      clearTimeout(this.timer.ref);
      await utils.timeout(800, this.timer);
    }

    this.setState({ loading: true });
    try {
      const res = await taskListApi.get(
        connection,
        page,
        rowsPerPage,
        sortedColumn,
        sortedOrder,
        filter
      );

      this.setState({
        rows: normalizeRows(res.payload),
        size: res.metadata.size,
        currentPage: page || 0,
        loading: false,
      });
      callback();
    } catch (error) {
      this.handleError(error);
      this.setState({ loading: false });
    }
  };

  closeNotification = () => {
    this.setState({ errorAlert: null });
  };

  handleError(err) {
    const { onError } = this.props;
    onError(err);
    this.setState({
      error: true,
      errorAlert: err,
    });
  }

  render() {
    const { loading, columns, rows, currentPage, size, error, errorAlert } = this.state;
    const { classes, lazyLoading } = this.props;

    let lazyLoadingProps;
    if (lazyLoading) {
      lazyLoadingProps = {
        currentPage,
        onChange: this.updateRows,
        size,
      };
    }

    return (
      <ThemeProvider theme={theme}>
        <Paper className={classes.paper}>
          {error ? (
            <ErrorComponent message={i18next.t('messages.errors.dataLoading')} />
          ) : (
            <Table
              loading={loading}
              title={i18next.t('table.title')}
              subtitle={i18next.t('table.subtitle')}
              columns={columns}
              rows={rows}
              rowsPerPageOptions={[10, 25, 50, 100]}
              lazyLoadingProps={lazyLoadingProps}
            />
          )}
        </Paper>
        <ErrorNotification message={errorAlert} onClose={this.closeNotification} />
      </ThemeProvider>
    );
  }
}

TaskList.propTypes = {
  classes: PropTypes.shape({
    paper: PropTypes.string,
  }),
  lazyLoading: PropTypes.bool,
  onError: PropTypes.func,
};

TaskList.defaultProps = {
  classes: {},
  lazyLoading: false,
  onError: () => {},
};

export default withStyles(styles)(TaskList);
