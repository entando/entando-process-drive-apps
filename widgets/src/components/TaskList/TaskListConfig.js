import React from 'react';
import { FormGroup, ControlLabel, Checkbox, HelpBlock, Row, Col } from 'patternfly-react';
import { SortableContainer, SortableElement } from 'react-sortable-hoc';
import arrayMove from 'array-move';
import { withStyles } from '@material-ui/core/styles';

import { getConnections } from 'api/pda/connections';
import { getGroups } from 'api/pda/groups';
import { getColumns } from 'api/pda/tasks';
import { normalizeConfigColumns, normalizeConfigGroups } from 'components/TaskList/normalizeData';
import taskListConfigType from 'types/taskListConfigType';
import RenderSwitch from 'components/common/RenderSwitch';
import Notification from 'components/common/Notification';

const generalOptions = [
  {
    key: 'newPageOnClick',
    label: 'Open new page on table row click',
    checked: false,
  },
  {
    key: 'showClaim',
    label: 'Show Claim Button',
    checked: true,
  },
  {
    key: 'showComplete',
    label: 'Show Complete Button',
    checked: true,
  },
];

const SortableRow = SortableElement(({ value, onColumnChange }) => (
  <tr>
    <td>
      <i className="fa fa-bars" style={{ cursor: 'ns-resize' }} /> {value.position}
    </td>
    <td>{value.name}</td>
    <td>
      <Checkbox
        style={{ margin: 0 }}
        bsClass="checkbox"
        checked={value.isVisible}
        onChange={onColumnChange}
      />
    </td>
    <td>
      <input
        type="text"
        className="form-control"
        onChange={onColumnChange}
        value={value.header || ''}
      />
    </td>
  </tr>
));

const SortableBody = SortableContainer(({ items, onColumnChange }) => (
  <tbody>
    {items.map((value, index) => (
      <SortableRow
        key={value.name}
        index={index}
        value={value}
        onColumnChange={onColumnChange(index)}
      />
    ))}
  </tbody>
));

const styles = {
  paper: {
    padding: 20,
  },
  title: {
    display: 'flex',
    alignItems: 'center',
    paddingBottom: 15,
  },
  subtitle: {
    padding: '20px 0 20px 0',
  },
  formControl: {
    minWidth: 200,
  },
};

class TaskListConfig extends React.Component {
  state = {
    config: {
      knowledgeSource: '',
      options: generalOptions,
      groups: [],
      columns: [],
    },
    sourceList: [],
    errorAlert: null,
  };

  componentDidMount = async () => {
    try {
      // get list of connections
      const sourceList = await getConnections();
      if (sourceList.errors && sourceList.errors.length) {
        throw sourceList.errors[0];
      }

      // fetch screen if a default config prop is passed
      this.setState({ sourceList: sourceList.payload }, this.fetchScreen);
    } catch (error) {
      this.handleError(error.message);
    }
  };

  componentDidUpdate = async prevProps => {
    const { config } = this.props;

    // refetch state if config changes
    if (JSON.stringify(config) !== JSON.stringify(prevProps.config)) {
      this.fetchScreen();
    }
  };

  fetchScreen = () => {
    const { config } = this.props;
    if (config && config.knowledgeSource) {
      this.onChangeSource(config.knowledgeSource, () => {
        this.setState({
          config: {
            ...config,
            groups: JSON.parse(config.groups),
            options: JSON.parse(config.options),
            columns: JSON.parse(config.columns),
          },
        });
      });
    }
  };

  onChangeSource = (e, cb) => {
    const { config } = this.state;
    const knowledgeSource = e.target ? e.target.value : e;
    this.setState(
      {
        config: {
          ...config,
          options: [...generalOptions],
          groups: [],
          columns: [],
          knowledgeSource,
        },
      },
      async () => {
        if (cb) {
          cb();
        } else {
          try {
            const { payload: groups } = await getGroups(knowledgeSource);
            const { payload: columns } = await getColumns(knowledgeSource);

            this.setState(state => ({
              config: {
                ...state.config,
                groups: normalizeConfigGroups(groups),
                columns: normalizeConfigColumns(columns),
              },
            }));
          } catch (error) {
            // eslint-disable-next-line no-console
            console.log('Error while trying to fetch groups and columns from PDA server', error);
          }
        }
      }
    );
  };

  handleSortStart = ({ node }) => {
    const tds = document.getElementsByClassName('table-sortableHelper')[0].childNodes;
    node.childNodes.forEach((n, idx) => {
      tds[idx].style.width = `${n.offsetWidth}px`;
    });
  };

  handleSortEnd = ({ oldIndex, newIndex }) => {
    this.setState(({ config }) => ({
      config: {
        ...config,
        columns: arrayMove(config.columns, oldIndex, newIndex).map((column, i) => ({
          ...column,
          position: i,
        })),
      },
    }));
  };

  handleColumnChange = index => e => {
    const { config } = this.state;
    const { value, checked, type } = e.target;
    const newColumns = [...config.columns];
    if (type === 'checkbox') {
      newColumns[index].isVisible = checked;
    } else {
      newColumns[index].header = value;
    }
    this.setState({ config: { ...config, columns: newColumns } });
  };

  handleOptions = index => e => {
    const { config } = this.state;
    const { value } = e.state;
    const newOptions = [...config.options];

    newOptions[index].checked = value;
    this.setState({ config: { ...config, options: newOptions } });
  };

  handleGroups = index => e => {
    const { config } = this.state;
    const { value } = e.state;
    const newGroups = [...config.groups];

    newGroups[index].checked = value;
    this.setState({ config: { ...config, groups: newGroups } });
  };

  closeNotification = () => {
    this.setState({ errorAlert: null });
  };

  handleError(err) {
    this.setState({
      errorAlert: err,
    });
  }

  render() {
    const { sourceList, config, errorAlert } = this.state;
    const { knowledgeSource, groups, columns, options } = config;
    return (
      <div>
        <Notification type="error" message={errorAlert} onClose={this.closeNotification} />
        <form>
          <Row>
            <Col xs={12}>
              <FormGroup controlId="connection">
                <ControlLabel>Knowledge Source</ControlLabel>
                <select
                  className="form-control"
                  value={knowledgeSource}
                  onChange={this.onChangeSource}
                >
                  <option value="">Select...</option>
                  {sourceList.map(source => (
                    <option key={source.name} value={source.name}>
                      {source.name}
                    </option>
                  ))}
                </select>
                <HelpBlock>Select one of the Kie server connections.</HelpBlock>
              </FormGroup>
            </Col>
          </Row>
          {knowledgeSource && (
            <section>
              <legend>General options</legend>
              <Row>
                <Col xs={12}>
                  {options.map((item, i) => (
                    <RenderSwitch
                      key={item.key}
                      id={item.key}
                      label={item.label}
                      checked={item.checked}
                      onChange={this.handleOptions(i)}
                    />
                  ))}
                </Col>
              </Row>
              <legend>BPM Groups</legend>
              <Row>
                <Col xs={12}>
                  {groups.map((item, i) => (
                    <RenderSwitch
                      key={item.key}
                      id={item.key}
                      label={item.label}
                      checked={item.checked}
                      onChange={this.handleGroups(i)}
                    />
                  ))}
                </Col>
              </Row>
              <legend>Task Columns</legend>
              <Row>
                <Col xs={12}>
                  <table className="table">
                    <thead>
                      <tr>
                        <th>Position</th>
                        <th>Column Name</th>
                        <th>Visible</th>
                        <th>Override Fields</th>
                      </tr>
                    </thead>
                    <SortableBody
                      items={columns}
                      helperClass="table-sortableHelper"
                      onSortStart={this.handleSortStart}
                      onSortEnd={this.handleSortEnd}
                      onColumnChange={this.handleColumnChange}
                    />
                  </table>
                </Col>
              </Row>
            </section>
          )}
        </form>
      </div>
    );
  }
}

TaskListConfig.propTypes = {
  config: taskListConfigType,
};

TaskListConfig.defaultProps = {
  config: {},
};

const TaskListConfigContainer = withStyles(styles)(TaskListConfig);

export default TaskListConfigContainer;
