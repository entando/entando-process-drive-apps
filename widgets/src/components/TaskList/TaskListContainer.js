import { MuiThemeProvider as ThemeProvider } from '@material-ui/core/styles';
import React from 'react';
import i18next from 'i18next';
import PropTypes from 'prop-types';
import Paper from '@material-ui/core/Paper';
import withStyles from '@material-ui/core/styles/withStyles';
import SVG from 'react-inlinesvg';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

import { DOMAINS, LOCAL } from 'api/constants';
import { getTasks } from 'api/pda/tasks';
import { getDiagram } from 'api/pda/processes';
import { getPageWidget } from 'api/app-builder/pages';
import utils from 'utils';

import { normalizeColumns, normalizeRows } from 'components/TaskList/normalizeData';
import SearchInput from 'components/common/SearchInput';
import Notification from 'components/common/Notification';
import ErrorComponent from 'components/common/ErrorComponent';
import Table from 'components/common/Table/Table';
import SimpleDialog from 'components/common/SimpleDialog';
import theme from 'theme';

const styles = {
  toolbar: {
    justifyContent: 'space-between',
    padding: '16px 16px 8px 16px',
    minHeight: 'unset',
  },
  noSubtitleToolbar: {
    padding: '8px 16px',
  },
  title: {
    textAlign: 'left',
  },
  paper: {
    minHeight: 459,
    position: 'relative',
  },
};

class TaskList extends React.Component {
  state = {
    loading: true,
    columns: [],
    rows: [],
    size: 0,
    connection: {},
    filter: '',
    blocker: '',
    errorAlert: null,
    lastPage: false,
    activeTab: 0,
    diagramModal: {
      open: false,
      title: '',
      body: '',
    },
    groups: [],
    page: 0,
  };

  timer = { ref: null };

  componentDidMount = async () => {
    const { lazyLoading, serviceUrl, pageCode, frameId, onSelectTask } = this.props;

    if (!LOCAL) {
      // set the PDA domain to the URL passed via props
      DOMAINS.PDA = serviceUrl;
    }

    try {
      // config will be fetched from app-builder
      const widgetConfigs = await getPageWidget(pageCode, frameId, 'TASK_LIST');
      if (widgetConfigs.errors && widgetConfigs.errors.length) {
        throw widgetConfigs.errors[0];
      }
      if (!widgetConfigs.payload) {
        throw new Error('messages.errors.widgetConfig');
      }

      const { config } = widgetConfigs.payload;
      const connection = config.knowledgeSource;
      const groups = JSON.parse(config.groups)
        .filter(group => group.checked)
        .map(group => group.key);

      const taskList = lazyLoading
        ? await getTasks({ connection, groups }, 0, 10)
        : await getTasks({ connection, groups });

      if (!taskList.payload) {
        throw new Error('messages.errors.errorResponse');
      }

      if (!taskList.payload.length) {
        this.setState({ blocker: 'taskList.emptyList' });
      } else {
        const options = JSON.parse(config.options);
        const rows = normalizeRows(taskList.payload);

        // dispatch onSelectTask event for the first item on list
        onSelectTask({ ...rows[0], pos: 0, groups });

        this.setState({
          loading: false,
          columns: normalizeColumns(JSON.parse(config.columns), rows[0], options, {
            openDiagram: this.openDiagram,
            selectTask: onSelectTask,
          }),
          rows,
          lastPage: taskList.metadata.lastPage === 1,
          size: taskList.metadata.size,
          connection: config.knowledgeSource,
          groups,
        });
      }
    } catch (error) {
      this.handleError(error.message, 'messages.errors.dataLoading');
    }
  };

  componentDidUpdate = prevProps => {
    const { lazyLoading } = this.props;
    if (prevProps.lazyLoading !== lazyLoading) {
      this.updateRows(lazyLoading ? 0 : undefined);
    }
  };

  updateRows = async (
    page,
    rowsPerPage = 10,
    sortedColumn,
    sortedOrder,
    filter,
    callback = () => {},
    withDelay
  ) => {
    const { connection, groups } = this.state;

    if (withDelay) {
      clearTimeout(this.timer.ref);
      await utils.timeout(800, this.timer);
    }

    this.setState({ loading: true });
    try {
      const res = await getTasks(
        { connection, groups },
        page,
        rowsPerPage,
        sortedColumn,
        sortedOrder,
        filter
      );
      if (!res.payload) {
        throw res.message;
      }

      this.setState({
        rows: normalizeRows(res.payload),
        size: res.metadata.size,
        lastPage: res.metadata.lastPage === 1,
        loading: false,
      });
      callback();
    } catch (error) {
      this.handleError(error, 'messages.errors.dataLoading');
      this.setState({ loading: false });
    }
  };

  openDiagram = row => async () => {
    const { connection } = this.state;

    if (row.processId && row.id) {
      const process = `${row.processId}@${row.id.split('@')[1]}`;
      this.setState({ loading: true });
      try {
        const diagram = await getDiagram(connection, process);
        this.setState({
          diagramModal: {
            title: `${i18next.t('taskList.diagramModalTitle')}: ${process}`,
            body: <SVG src={diagram} />,
            open: true,
          },
        });
      } catch (error) {
        this.handleError(error);
      } finally {
        this.setState({ loading: false });
      }
    }
  };

  onCloseDiagramModal = () => {
    const { diagramModal } = this.state;
    this.setState({ diagramModal: { ...diagramModal, open: false } });
  };

  closeNotification = () => {
    this.setState({ errorAlert: null });
  };

  handleChangeFilter = event => {
    const { lazyLoading } = this.props;
    const { rowsPerPage, sortedColumn, sortOrder } = this.state;
    const filter = event.target.value;

    this.setState({ filter });
    if (lazyLoading) {
      this.updateRows(0, rowsPerPage, sortedColumn, sortOrder, filter, undefined, true);
    }
  };

  handleChangeTab = (_, activeTab) => {
    this.setState({ activeTab });
  };

  handleChangePage = page => {
    this.setState({
      page,
    });
  };

  handleRowSelect = (row, idx) => {
    const { onSelectTask } = this.props;
    const { page, rowsPerPage } = this.state;
    const pos = page * (rowsPerPage || 10) + idx;
    onSelectTask({ ...row, pos });
  };

  handleError(err, blocker = '') {
    const { onError } = this.props;
    onError(err);
    this.setState({
      errorAlert: err.toString(),
      blocker,
    });
  }

  render() {
    const {
      loading,
      columns,
      rows,
      size,
      blocker,
      errorAlert,
      lastPage,
      diagramModal,
      filter,
      activeTab,
    } = this.state;
    const { classes, lazyLoading } = this.props;

    let lazyLoadingProps;
    if (lazyLoading) {
      lazyLoadingProps = {
        onChange: this.updateRows,
        size,
        lastPage,
      };
    }

    return (
      <ThemeProvider theme={theme}>
        <Paper className={classes.paper}>
          {blocker ? (
            <ErrorComponent message={blocker} />
          ) : (
            <>
              <Toolbar className={classes.toolbar}>
                <div className={classes.title}>
                  <Typography variant="h2">{i18next.t('table.title')}</Typography>
                </div>
                <div>
                  <SearchInput value={filter} onChange={this.handleChangeFilter} />
                </div>
              </Toolbar>
              <Tabs
                indicatorColor="primary"
                textColor="primary"
                onChange={this.handleChangeTab}
                value={activeTab}
              >
                <Tab label={i18next.t('taskList.tabs.myTasks')} />
              </Tabs>
              <Table
                loading={loading}
                columns={columns}
                rows={rows}
                rowsPerPageOptions={[10, 25, 50, 100]}
                lazyLoadingProps={lazyLoadingProps}
                onRowClick={this.handleRowSelect}
                onChangePage={this.handleChangePage}
              />
            </>
          )}
        </Paper>
        <SimpleDialog
          open={diagramModal.open}
          title={diagramModal.title}
          body={diagramModal.body}
          onClose={this.onCloseDiagramModal}
          maxWidth="xl"
          fullWidth
        />
        <Notification type="error" message={errorAlert} onClose={this.closeNotification} />
      </ThemeProvider>
    );
  }
}

TaskList.propTypes = {
  classes: PropTypes.shape({
    paper: PropTypes.string,
    toolbar: PropTypes.string,
    title: PropTypes.string,
  }),
  lazyLoading: PropTypes.bool,
  onError: PropTypes.func,
  onSelectTask: PropTypes.func,
  serviceUrl: PropTypes.string,
  pageCode: PropTypes.string,
  frameId: PropTypes.string,
};

TaskList.defaultProps = {
  classes: {},
  lazyLoading: true,
  onError: () => {},
  onSelectTask: () => {},
  serviceUrl: '',
  pageCode: '',
  frameId: '',
};

export default withStyles(styles)(TaskList);
