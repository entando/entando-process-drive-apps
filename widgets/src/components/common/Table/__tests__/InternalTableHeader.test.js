import React from 'react';
import { render } from '@testing-library/react';

import InternalTableHeader from 'components/common/Table/InternalTableHeader';
import WIDGET_CONFIGS from 'mocks/app-builder/pages';
import { normalizeColumns } from 'components/TaskList/normalizeData';
import jsonRows from 'mocks/pda/tasks.json';

const columns = normalizeColumns(
  JSON.parse(WIDGET_CONFIGS.TASK_LIST.payload.config.columns),
  jsonRows.payload[0],
  JSON.parse(WIDGET_CONFIGS.TASK_LIST.payload.config.options)
);

describe('<InternalTableHeader />', () => {
  it('renders snapshot correctly', () => {
    const { container } = render(
      <table>
        <thead>
          <tr>
            <InternalTableHeader column={columns[2]} sortOrder="asc" />
          </tr>
        </thead>
      </table>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders snapshot correctly without sort functionality', () => {
    const { container } = render(
      <table>
        <thead>
          <tr>
            <InternalTableHeader column={columns[0]} sortOrder="asc" />
          </tr>
        </thead>
      </table>
    );

    expect(container).toMatchSnapshot();
  });
});
