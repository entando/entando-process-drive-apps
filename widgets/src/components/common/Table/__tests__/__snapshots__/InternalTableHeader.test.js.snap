// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InternalTableHeader /> renders snapshot correctly 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <td
          class="MuiTableCell-root MuiTableCell-head"
          style="white-space: nowrap;"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiTableSortLabel-root TableSortLabel-notActive-15"
            role="button"
            tabindex="0"
          >
            createdBy
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </td>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`<InternalTableHeader /> renders snapshot correctly without sort functionality 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <td
          class="MuiTableCell-root MuiTableCell-head"
          style="white-space: nowrap;"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiTableSortLabel-root TableSortLabel-notActive-47"
            role="button"
            tabindex="0"
          >
            id
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </td>
      </tr>
    </thead>
  </table>
</div>
`;
